src/backend/distributed/planner/recursive_planning.c

```diff
@@ -252,10 +309,516 @@ RecursivelyPlanSubqueriesAndCTEs(Query *query, RecursivePlanningContext *context
 		RecursivelyPlanAllSubqueries((Node *) query->jointree->quals, context);
 	}
 
+	/*
+	 * If the query doesn't have distribution key equality,
+	 * recursively plan some of its subqueries.
+	 */
+	if (ShouldRecursivelyPlanNonColocatedSubqueries(query, context))
+	{
+		RecursivelyPlanNonColocatedSubqueries(query, context);
+	}
+
+	return NULL;
+}
+
+
+/*
+ * ShouldRecursivelyPlanNonColocatedSubqueries returns true if the input query contains joins
+ * that are not on the distribution key.
+ * *
+ * Note that at the point that this function is called, we've already recursively planned all
+ * the leaf subqueries. Thus, we're actually checking whether the joins among the subqueries
+ * on the distribution key or not.
+ */
+static bool
+ShouldRecursivelyPlanNonColocatedSubqueries(Query *subquery,
+											RecursivePlanningContext *context)
+{
+	/*
+	 * If the input query already contains the equality, simply return since it is not
+	 * possible to find any non colocated subqueries.
+	 */
+	if (context->queryContainsDistributionKeyEquality)
+	{
+		return false;
+	}
+
+	/*
+	 * This check helps us in two ways:
+	 *   (i) We're not targeting queries that don't include subqueries at all,
+	 *       they should go through regular planning.
+	 *  (ii) Lower level subqueries are already recursively planned, so we should
+	 *       only bother non-colocated subquery joins, which only happens when
+	 *       there are subqueries.
+	 */
+	if (SubqueryEntryList(subquery) == NIL && SublinkList(subquery) == NIL)
```

Comment:
Since we're immediately throwing away these lists and this function may get called very often, could we implement this in a way that does not allocate memory?

Reply:
Sure, I can write new functions which returns immediately when it finds a sublink or subquery. But, that'd mean to almost copy both functions. 

I'm not that much concerned about the memory given that  I expect the other parts of the planning would be the bottleneck not here if the there are many subqueries in the query (i.e., the lists are lengthy).

Still, I could implement more efficient functions I you prefer to do so.

Reply:
Kind of. Also for readability, it'd be much nicer if there was a `ContainsSubquery(subquery)` function.